bpf: Expand LPM key for future usage.

joestringer opened the pull request:
Expand the lpm keys so that in future we can place port/protocol inside for supporting CIDR-dependent L4, without breaking the map format again.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/3173)
<!-- Reviewable:end -->

joestringer committed:
bpf: Expand LPM key,value for future usage.

Expand the lpm key, value so that in future we can place port/protocol
and proxyport inside inside for supporting CIDR-dependent L4.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon reviewed (commented):
How will this affect upgrade?

joestringer committed:
bpf: Reuse lpm trie keys in bpf_xdp

bpf_xdp previously defined its own CIDR key/value sizes and types,
however in userspace we use one set of definitions for all CIDR maps.
Share the definition so that any update to one of these will apply to
all of them.

Signed-off-by: Joe Stringer <joe@covalent.io>

borkmann commented on bpf/lib/maps.h:
```diff
@@ -121,18 +121,22 @@ struct bpf_elf_map __section_maps tunnel_endpoint_map = {
 struct bpf_lpm_trie_key6 {
 	struct bpf_lpm_trie_key lpm_key;
+	__u16 pad1;
+	__u16 pad2;
```
Why not making this __u32?

joestringer replied:
Originally I was thinking in terms of port + proto, which would need to be split into (at least) two fields anyway. I can fix this up for simplicity.

borkmann commented on bpf/bpf_xdp.c:
```diff
@@ -107,6 +107,7 @@ static __always_inline int check_v4(struct xdp_md *xdp)
 	__builtin_memcpy(pfx.lpm.data, &ipv4_hdr->saddr, sizeof(pfx.addr));
 	pfx.lpm.prefixlen = 32;
+	pfx.pad = 0;
 
```
I think here may be a bug: in the `lpm_v{4,6}_key` case you also add this in front of the key, therefore prefixlen needs to grow as well by as in the other lpm case due to the extra zeroed prefix.

borkmann commented on bpf/lib/maps.h:
```diff
@@ -121,18 +121,22 @@ struct bpf_elf_map __section_maps tunnel_endpoint_map = {
 					   union v6addr *addr, __u32 prefix)
 {
-	struct bpf_lpm_trie_key6 key = { { prefix }, *addr };
+	struct bpf_lpm_trie_key6 key = { { prefix + sizeof(__u32) }, 0, 0, *addr };
```
Shouldn't this be: sizeof(__u32) * 8 ? Call-sites use prefix value of either 32 or 128, and if we extend it by 4 bytes, it should sum up to either 64 or 160 eventually, no?

joestringer replied:
right, I just thought about this last night and apparently didn't think clearly enough. I'll fix it.

borkmann reviewed (approved):
LGTM
